File: backend/app/schemas/expansion.py
"""
扩写模块 Pydantic Schemas
"""
from datetime import datetime
from typing import Any, Dict, List, Literal, Optional

from pydantic import BaseModel, ConfigDict, Field, field_validator

# --- 类型常量 ---

VALID_EXPANSION_LEVELS = Literal["light", "medium", "deep"]

VALID_SOURCE_TYPES = Literal["upload", "novel", "drama", "manual"]

VALID_PROJECT_STATUSES = Literal[
    "created", "analyzed", "segmented", "expanding", "paused", "error", "completed"
]

VALID_SEGMENT_STATUSES = Literal["pending", "expanding", "completed", "error", "skipped"]

VALID_EXECUTION_MODES = Literal["auto", "step_by_step"]

VALID_CONVERT_TARGETS = Literal["novel", "drama"]


# --- 文风画像 ---

class StyleProfile(BaseModel):
    """文风画像"""
    narrative_pov: Optional[str] = Field(None, description="叙事视角")
    tone: Optional[str] = Field(None, description="基调/氛围")
    sentence_style: Optional[str] = Field(None, description="句式风格")
    vocabulary: Optional[str] = Field(None, description="词汇特点")
    rhythm: Optional[str] = Field(None, description="节奏特点")
    notable_features: Optional[str] = Field(None, description="显著特征")


# --- AI 配置（复用 drama 模式）---

class ExpansionAIPromptConfig(BaseModel):
    """AI 提示词配置"""
    system_prompt: Optional[str] = Field(None, description="系统提示词")
    temperature: Optional[float] = Field(None, ge=0.0, le=2.0, description="温度参数")
    max_tokens: Optional[int] = Field(None, gt=0, description="最大输出 token 数")


class ExpansionAIConfig(BaseModel):
    """AI 配置"""
    provider: Optional[str] = Field(None, description="AI 提供商")
    model: Optional[str] = Field(None, description="模型名称")
    prompt_config: Optional[ExpansionAIPromptConfig] = Field(None, description="提示词配置")


# --- ExpansionProject Schemas ---

class ExpansionProjectCreate(BaseModel):
    """创建扩写项目请求"""
    title: str = Field(..., min_length=1, max_length=200, description="项目标题")
    source_type: VALID_SOURCE_TYPES = Field(..., description="来源类型: upload/novel/drama/manual")
    original_text: str = Field(..., max_length=30000, description="原始文本，最多30000字")
    source_ref: Optional[Dict[str, Any]] = Field(None, description="来源引用")
    expansion_level: VALID_EXPANSION_LEVELS = Field("medium", description="扩写深度")
    target_word_count: Optional[int] = Field(None, gt=0, description="目标字数")
    style_instructions: Optional[str] = Field(None, description="文风调整指令")
    ai_config: Optional[ExpansionAIConfig] = Field(None, description="AI配置")
    execution_mode: VALID_EXECUTION_MODES = Field("auto", description="执行模式")
    metadata_: Optional[Dict[str, Any]] = Field(None, alias="metadata", description="元数据")

    model_config = ConfigDict(populate_by_name=True)

    @field_validator("original_text")
    @classmethod
    def validate_text_length(cls, v: str) -> str:
        if len(v) > 30000:
            raise ValueError("原文超过30000字限制")
        return v


class ExpansionProjectUpdate(BaseModel):
    """更新扩写项目请求"""
    title: Optional[str] = Field(None, min_length=1, max_length=200, description="项目标题")
    expansion_level: Optional[VALID_EXPANSION_LEVELS] = Field(None, description="扩写深度")
    target_word_count: Optional[int] = Field(None, gt=0, description="目标字数")
    style_instructions: Optional[str] = Field(None, description="文风调整指令")
    status: Optional[VALID_PROJECT_STATUSES] = Field(None, description="状态")
    execution_mode: Optional[VALID_EXECUTION_MODES] = Field(None, description="执行模式")
    ai_config: Optional[ExpansionAIConfig] = Field(None, description="AI配置")
    metadata_: Optional[Dict[str, Any]] = Field(None, alias="metadata", description="元数据")

    model_config = ConfigDict(populate_by_name=True)


class ExpansionProjectResponse(BaseModel):
    """扩写项目响应"""
    model_config = ConfigDict(from_attributes=True, populate_by_name=True)

    id: int
    user_id: int
    title: str
    source_type: str
    source_ref: Optional[Dict[str, Any]] = None
    original_text: Optional[str] = None
    word_count: int
    summary: Optional[str] = None
    style_profile: Optional[Dict[str, Any]] = None
    expansion_level: str
    target_word_count: Optional[int] = None
    style_instructions: Optional[str] = None
    ai_config: Optional[Dict[str, Any]] = None
    status: str
    execution_mode: str
    version: int
    metadata_: Optional[Dict[str, Any]] = Field(
        None,
        serialization_alias="metadata",
        description="元数据"
    )
    created_at: datetime
    updated_at: Optional[datetime] = None


class ExpansionProjectListResponse(BaseModel):
    """扩写项目列表响应（不含original_text）"""
    items: List["ExpansionProjectListItem"]
    total: int
    page: int
    page_size: int


class ExpansionProjectListItem(BaseModel):
    """扩写项目列表项"""
    model_config = ConfigDict(from_attributes=True)

    id: int
    user_id: int
    title: str
    source_type: str
    word_count: int
    status: str
    expansion_level: str
    created_at: datetime
    updated_at: Optional[datetime] = None


# --- ExpansionSegment Schemas ---

class ExpansionSegmentResponse(BaseModel):
    """扩写分段响应"""
    model_config = ConfigDict(from_attributes=True)

    id: int
    project_id: int
    sort_order: int
    title: Optional[str] = None
    original_content: str
    expanded_content: Optional[str] = None
    expansion_level: Optional[str] = None
    custom_instructions: Optional[str] = None
    status: str
    error_message: Optional[str] = None
    original_word_count: int
    expanded_word_count: Optional[int] = None
    created_at: datetime
    updated_at: Optional[datetime] = None


class ExpansionSegmentUpdate(BaseModel):
    """更新扩写分段请求"""
    title: Optional[str] = Field(None, max_length=200, description="段落标题")
    expansion_level: Optional[VALID_EXPANSION_LEVELS] = Field(None, description="扩写深度")
    custom_instructions: Optional[str] = Field(None, description="特殊扩写指令")

    model_config = ConfigDict(populate_by_name=True)


# --- 分段操作请求 ---

class SegmentSplitRequest(BaseModel):
    """分段拆分请求"""
    segment_id: int = Field(..., description="要拆分的分段ID")
    split_position: int = Field(..., ge=0, description="拆分位置（字符索引）")


class SegmentMergeRequest(BaseModel):
    """分段合并请求"""
    segment_ids: List[int] = Field(..., min_length=2, description="要合并的分段ID列表")

    @field_validator("segment_ids")
    @classmethod
    def validate_segment_ids(cls, v: List[int]) -> List[int]:
        if len(v) < 2:
            raise ValueError("至少需要2个分段才能合并")
        return v


class SegmentReorderRequest(BaseModel):
    """分段重排序请求"""
    segment_ids: List[int] = Field(..., description="按新顺序排列的分段ID列表")


# --- 导入请求 ---

class ImportFromNovelRequest(BaseModel):
    """从小说导入请求"""
    title: str = Field(..., min_length=1, max_length=200, description="项目标题")
    project_id: int = Field(..., description="小说项目ID")
    chapter_ids: Optional[List[int]] = Field(None, description="章节ID列表，不指定则导入全部")
    expansion_level: VALID_EXPANSION_LEVELS = Field("medium", description="扩写深度")
    ai_config: Optional[ExpansionAIConfig] = Field(None, description="AI配置")
    execution_mode: VALID_EXECUTION_MODES = Field("auto", description="执行模式")


class ImportFromDramaRequest(BaseModel):
    """从剧本导入请求"""
    title: str = Field(..., min_length=1, max_length=200, description="项目标题")
    project_id: int = Field(..., description="剧本项目ID")
    expansion_level: VALID_EXPANSION_LEVELS = Field("medium", description="扩写深度")
    ai_config: Optional[ExpansionAIConfig] = Field(None, description="AI配置")
    execution_mode: VALID_EXECUTION_MODES = Field("auto", description="执行模式")


# --- 转换请求 ---

class ConvertRequest(BaseModel):
    """转换请求"""
    target: VALID_CONVERT_TARGETS = Field(..., description="目标类型: novel/drama")


# --- 扩写操作请求 ---

class ExpandSegmentRequest(BaseModel):
    """单段扩写请求"""
    segment_id: int = Field(..., description="要扩写的分段ID")
    instructions: Optional[str] = Field(None, description="额外指令")

File: frontend/src/api/expansion.ts
/**
 * 扩写模块 API
 * 支持 REST CRUD 和 SSE 流式调用
 */
import request from './request'
import { getAccessToken } from './request'

// ── Types ──

export type ProjectStatus = 'created' | 'analyzed' | 'segmented' | 'expanding' | 'paused' | 'error' | 'completed'
export type SegmentStatus = 'pending' | 'expanding' | 'completed' | 'error' | 'skipped'
export type ExpansionLevel = 'light' | 'medium' | 'deep'
export type SourceType = 'upload' | 'novel' | 'drama' | 'manual'
export type ExecutionMode = 'auto' | 'step_by_step'
export type ConvertTarget = 'novel' | 'drama'

export interface ExpansionAIConfig {
  provider?: string
  model?: string
  temperature?: number
  max_tokens?: number
}

export interface ExpansionProject {
  id: number
  user_id: number
  title: string
  source_type: 'upload' | 'novel' | 'drama' | 'manual'
  word_count: number
  summary: string | null
  style_profile: Record<string, unknown> | null
  expansion_level: 'light' | 'medium' | 'deep'
  target_word_count: number | null
  style_instructions: string | null
  status: 'created' | 'analyzed' | 'segmented' | 'expanding' | 'paused' | 'error' | 'completed'
  execution_mode: 'auto' | 'step_by_step'
  version: number
  created_at: string
  updated_at: string | null
}

export interface ExpansionProjectListItem {
  id: number
  title: string
  source_type: 'upload' | 'novel' | 'drama' | 'manual'
  word_count: number
  expansion_level: 'light' | 'medium' | 'deep'
  status: string
  created_at: string
}

export interface ExpansionSegment {
  id: number
  project_id: number
  sort_order: number
  title: string | null
  original_content: string
  expanded_content: string | null
  expansion_level: 'light' | 'medium' | 'deep' | null
  custom_instructions: string | null
  status: 'pending' | 'expanding' | 'completed' | 'error' | 'skipped'
  error_message: string | null
  original_word_count: number
  expanded_word_count: number | null
  created_at: string
  updated_at: string | null
}

export interface CreateExpansionProjectData {
  title: string
  source_type: 'upload' | 'novel' | 'drama' | 'manual'
  original_text: string
  expansion_level?: 'light' | 'medium' | 'deep'
  target_word_count?: number
  style_instructions?: string
  execution_mode?: 'auto' | 'step_by_step'
  ai_config?: ExpansionAIConfig
}

export interface UpdateExpansionProjectData {
  title?: string
  expansion_level?: 'light' | 'medium' | 'deep'
  target_word_count?: number
  style_instructions?: string
  execution_mode?: 'auto' | 'step_by_step'
  ai_config?: ExpansionAIConfig
  summary?: string
}

export interface ImportFromNovelData {
  project_id: number
  chapter_ids: number[]
  title?: string
}

export interface ImportFromDramaData {
  project_id: number
  title?: string
}

export interface SegmentSplitData {
  segment_id: number
  split_position: number
}

export interface SegmentMergeData {
  segment_ids: number[]
}

export interface ConvertData {
  target: 'novel' | 'drama'
}

export interface UpdateSegmentData {
  title?: string
  expansion_level?: 'light' | 'medium' | 'deep'
  custom_instructions?: string
}

// ── SSE Stream Callbacks ──

export interface StreamCallbacks {
  onText?: (text: string) => void
  onEvent?: (type: string, data: unknown) => void
  onDone?: (data?: unknown) => void
  onError?: (message: string) => void
}

// ── Project API ──

export async function getExpansionProjects(params?: {
  status?: string
  expansion_level?: string
  page?: number
  page_size?: number
}): Promise<ExpansionProjectListItem[]> {
  const res = await request.get<{ items: ExpansionProjectListItem[] }>('/expansion/', { params })
  return res.items
}

export async function createExpansionProject(data: CreateExpansionProjectData): Promise<ExpansionProject> {
  return request.post<ExpansionProject>('/expansion/', data)
}

export async function uploadExpansionProject(
  file: File,
  title?: string,
  expansion_level?: 'light' | 'medium' | 'deep',
  target_word_count?: number,
  style_instructions?: string,
  execution_mode?: 'auto' | 'step_by_step',
): Promise<ExpansionProject> {
  const formData = new FormData()
  formData.append('file', file)
  if (title) formData.append('title', title)
  if (expansion_level) formData.append('expansion_level', expansion_level)
  if (target_word_count) formData.append('target_word_count', String(target_word_count))
  if (style_instructions) formData.append('style_instructions', style_instructions)
  if (execution_mode) formData.append('execution_mode', execution_mode)

  return request.post<ExpansionProject>('/expansion/upload', formData, {
    headers: { 'Content-Type': 'multipart/form-data' },
  })
}

export async function importFromNovel(data: ImportFromNovelData): Promise<ExpansionProject> {
  return request.post<ExpansionProject>('/expansion/import', { source: 'novel', ...data })
}

export async function importFromDrama(data: ImportFromDramaData): Promise<ExpansionProject> {
  return request.post<ExpansionProject>('/expansion/import', { source: 'drama', ...data })
}

export async function getExpansionProject(id: number): Promise<ExpansionProject> {
  return request.get<ExpansionProject>(`/expansion/${id}`)
}

export async function updateExpansionProject(id: number, data: UpdateExpansionProjectData): Promise<ExpansionProject> {
  return request.put<ExpansionProject>(`/expansion/${id}`, data)
}

export async function deleteExpansionProject(id: number): Promise<void> {
  return request.delete(`/expansion/${id}`)
}

// ── Analysis (SSE) ──

export function streamAnalyzeProject(id: number, callbacks: StreamCallbacks): AbortController {
  return _expansionStreamRequest(`/api/v1/expansion/${id}/analyze`, {}, callbacks)
}

export function streamResegmentProject(id: number, callbacks: StreamCallbacks): AbortController {
  return _expansionStreamRequest(`/api/v1/expansion/${id}/resegment`, {}, callbacks)
}

// ── Segments ──

export async function getSegments(projectId: number): Promise<ExpansionSegment[]> {
  return request.get<ExpansionSegment[]>(`/expansion/${projectId}/segments`)
}

export async function updateSegment(
  projectId: number,
  segId: number,
  data: UpdateSegmentData,
): Promise<ExpansionSegment> {
  return request.put<ExpansionSegment>(`/expansion/${projectId}/segments/${segId}`, data)
}

export async function splitSegment(
  projectId: number,
  data: SegmentSplitData,
): Promise<ExpansionSegment[]> {
  return request.post<ExpansionSegment[]>(`/expansion/${projectId}/segments/split`, data)
}

export async function mergeSegments(
  projectId: number,
  data: SegmentMergeData,
): Promise<ExpansionSegment> {
  return request.post<ExpansionSegment>(`/expansion/${projectId}/segments/merge`, data)
}

export async function reorderSegments(projectId: number, order: number[]): Promise<void> {
  return request.put(`/expansion/${projectId}/segments/reorder`, { segment_ids: order })
}

// ── Expansion (SSE) ──

export function streamExpandProject(id: number, callbacks: StreamCallbacks): AbortController {
  return _expansionStreamRequest(`/api/v1/expansion/${id}/expand`, {}, callbacks)
}

export function streamExpandSegment(
  projectId: number,
  segId: number,
  callbacks: StreamCallbacks,
  data?: { expansion_level?: 'light' | 'medium' | 'deep'; custom_instructions?: string },
): AbortController {
  return _expansionStreamRequest(
    `/api/v1/expansion/${projectId}/segments/${segId}/expand`,
    data || {},
    callbacks,
  )
}

export async function pauseExpansion(id: number): Promise<void> {
  return request.post(`/expansion/${id}/pause`)
}

export function streamResumeExpansion(id: number, callbacks: StreamCallbacks): AbortController {
  return _expansionStreamRequest(`/api/v1/expansion/${id}/resume`, {}, callbacks)
}

export function streamRetrySegment(
  projectId: number,
  segId: number,
  callbacks: StreamCallbacks,
): AbortController {
  return _expansionStreamRequest(
    `/api/v1/expansion/${projectId}/segments/${segId}/retry`,
    {},
    callbacks,
  )
}

export function streamResegmentSegments(
  projectId: number,
  segmentIds: number[],
  callbacks: StreamCallbacks,
): AbortController {
  return _expansionStreamRequest(
    `/api/v1/expansion/${projectId}/segments/resegment`,
    { segment_ids: segmentIds },
    callbacks,
  )
}

// ── Export & Convert ──

export function getExportUrl(
  id: number,
  format: 'txt' | 'md' | 'docx',
  version: 'original' | 'expanded' | 'both',
): string {
  return `/api/v1/expansion/${id}/export?format=${format}&version=${version}`
}

export async function convertProject(
  id: number,
  data: ConvertData,
): Promise<{ project_id: number; project_type: string }> {
  return request.post(`/expansion/${id}/convert`, data)
}

// ── Internal SSE helper ──

function _expansionStreamRequest(
  url: string,
  body: Record<string, unknown>,
  callbacks: StreamCallbacks,
  timeoutMs: number = 600000, // 默认 10 分钟超时（长文本分析可能需要）
): AbortController {
  const controller = new AbortController()
  const headers: Record<string, string> = { 'Content-Type': 'application/json' }
  const token = getAccessToken()
  if (token) {
    headers['Authorization'] = `Bearer ${token}`
  }

  // 设置整体超时
  const timeoutId = setTimeout(() => {
    controller.abort()
    callbacks.onError?.('请求超时，文本过长或服务器响应过慢')
  }, timeoutMs)

  fetch(url, {
    method: 'POST',
    headers,
    body: JSON.stringify(body),
    signal: controller.signal,
    credentials: 'same-origin',
  })
    .then(async (response) => {
      console.log('[SSE] Response status:', response.status, 'headers:', Array.from(response.headers.entries()))
      if (!response.ok) {
        const err = await response.json().catch(() => ({ detail: '请求失败' }))
        callbacks.onError?.(err.detail || '请求失败')
        return
      }

      const reader = response.body?.getReader()
      if (!reader) {
        callbacks.onError?.('无法读取响应流')
        return
      }
      console.log('[SSE] Reader acquired, starting to read stream...')

      const decoder = new TextDecoder()
      let buffer = ''
      // TTFB 超时：首字节最多等 60 秒
      const TTFB_TIMEOUT = 60000
      // 读取超时：后续每块数据最多等 180 秒
      const READ_TIMEOUT = 180000

      let readTimeoutId: ReturnType<typeof setTimeout> | null = null
      let isFirstChunk = true

      const resetReadTimeout = () => {
        if (readTimeoutId) clearTimeout(readTimeoutId)
        readTimeoutId = setTimeout(() => {
          controller.abort()
          callbacks.onError?.('服务器响应超时，AI 模型可能过慢或网络不稳定')
        }, READ_TIMEOUT)
      }

      // 首字节超时检测
      const ttfbTimeoutId = setTimeout(() => {
        if (isFirstChunk) {
          controller.abort()
          callbacks.onError?.('AI 响应超时（60 秒未收到首字节），请重试或检查 AI 服务状态')
        }
      }, TTFB_TIMEOUT)

      resetReadTimeout()

      while (true) {
        const { done, value } = await reader.read()
        console.log('[SSE] Read chunk:', { done, valueLength: value?.length })
        if (done) break

        if (isFirstChunk) {
          clearTimeout(ttfbTimeoutId)
          isFirstChunk = false
        }

        resetReadTimeout()

        buffer += decoder.decode(value, { stream: true })
        const lines = buffer.split('\n')
        buffer = lines.pop() || ''

        for (const line of lines) {
          if (line.startsWith('data: ')) {
            try {
              const payload = JSON.parse(line.slice(6))
              const { type, text, ...rest } = payload

              if (text !== undefined) {
                callbacks.onText?.(text)
              }

              if (type === 'done') {
                if (readTimeoutId) clearTimeout(readTimeoutId)
                callbacks.onDone?.(rest)
                return
              }

              if (type === 'error') {
                if (readTimeoutId) clearTimeout(readTimeoutId)
                callbacks.onError?.(payload.message || '发生错误')
                return
              }

              // Handle all other event types: status, segments, segment_start, segment_done, await_confirm, phase
              if (type) {
                callbacks.onEvent?.(type, payload)
              }
            } catch {
              // ignore parse errors
            }
          }
        }
      }
      if (readTimeoutId) clearTimeout(readTimeoutId)
      if (ttfbTimeoutId) clearTimeout(ttfbTimeoutId)
      callbacks.onDone?.()
    })
    .catch((err) => {
      console.error('[SSE] Fetch error:', err)
      if (err.name !== 'AbortError') {
        callbacks.onError?.(err.message || '网络请求失败')
      }
    })
    .finally(() => {
      clearTimeout(timeoutId)
    })

  return controller
}